### PR TITLE
Implemented fns for direct sum and kronecker product of matrices

### PR DESCRIPTION
**Pull request by MagusMachinae**
Hey Sam, figured I should finally get around to doing these. Haven't written up tests yet, because I'm not sure if this is the best representation just yet, but I can write up a couple over the next couple days as this gets tidied up.

### CODE REVIEW

**Comment by sritchie** [comment]
This is excellent!! I’m back in town tomorrow and can take a close look then. Thank you, @MagusMachinae !

**Comment by sritchie** [comment]
@MagusMachinae nice! Would you mind adding some tests? That will be great, to see some expected results and any edge cases you are aware of. Then I can help get generative tests in and docs if we want those.

**Comment by sritchie** [comment]
Ping on this, @MagusMachinae !

**Comment by MagusMachinae** [comment]
Heya, sorry I've been out of the loop a bit, haven't been feeling well.

I'll get the tests in tonight my time.

**Comment by MagusMachinae** [comment]
There isn't much in the way of edge cases that I could come up with, there aren't any dimensionality conditions imposed by either of these functions. The only issue I saw as an edge case would be the notion that a block matrix produced by ⊗ is equivalent to an unpartitioned matrix with the same values in it, but that's more of a UX/ambiguous shorthand thing it feels.

File: src/sicmutils/matrix.cljc [hunk 1]
```diff
@@ -861,6 +861,32 @@
...
+  (fmap (fn [x] (g/* x matrix-B)) matrix-A))
+
+(defn dim-sum
+  "Initial implementation of Direct Sum"
+  [matrix-A matrix-B]
+  (let [new-row-count (+ (num-rows matrix-A) (num-rows matrix-B))
+        new-col-count (+ (num-cols matrix-A) (num-cols matrix-B))
+        B-row-offset  (- new-row-count (num-rows matrix-B))
+        B-col-offset  (- new-col-count (num-cols matrix-B))]
+    (fmap (fn [x] (if (nil? x)
```
[the remark this answers is not in the document]
Comment (MagusMachinae) [reply src/sicmutils/matrix.cljc:877]: By providing zero as the last argument to get-in?

File: test/sicmutils/matrix_test.cljc [hunk 2]
```diff
@@ -874,3 +874,27 @@
...
                        (s/down 30 -180 180))
                (g/divide H)))))))
+
+(deftest kronecker-product
+  (testing "Kronecker product, ⊗, of two matrices producing Block Matrix"
+    (let [A (m/by-rows [1 3 2]
+                       [2 3 1])
+          B (m/by-rows [1 6]
+                       [0 1])]
+      (is (= (m/matrix->vector
```
[the remark this answers is not in the document]
Comment (MagusMachinae) [reply test/sicmutils/matrix_test.cljc:884]: I was using matrix->vector because I wasn't sure whether the result of Kronecker products should be an actual block matrix, to be explicit: nested, or the fully simplified forms on the far right shown here.

image

I wasn't quite sure what the expected result would be, and was under the assumption that I would need to cast to a vector to call `flatten` on it.
I also was using it to suppress type tags on the matrix objects in the REPL, which got pretty out of hand when trying to replacing every element of a matrix with a matrix.

File: src/sicmutils/matrix.cljc [hunk 3]
```diff
@@ -861,6 +861,32 @@
     (when-not (= r c) (u/illegal "not square"))
     (determinant (g/- (g/* x (I r)) m))))
 
+(defn kronecker-product
+  "Computes a block matrix by mapping over the left matrix, multiplying elementwise
+by the right matrix."
+  [matrix-A matrix-B]
+  (fmap (fn [x] (g/* x matrix-B)) matrix-A))
+
+(defn dim-sum
```
[the remark this answers is not in the document]
Comment (MagusMachinae) [reply src/sicmutils/matrix.cljc:870]: I was having trouble coming up with an explanation of what a direct sum of matrices actually is without leaning too much on block matrices. I felt like it should explain how the matrices are being combined by seeing the terms of the left one as offset members and combining them into one space, but wasn't sure on the wording. QM stuff has me thinking about sums of Hilbert spaces, which is too specific, but I was trying to kick off of that thought to explain what's going on.

**Comment by sritchie** [comment]
Whoops, I did not mean to delete this, @MagusMachinae . I see that I did it accidentally by deleting the branch that this was merged into. If you want to re-open it can you make a PR against `main`? Otherwise I can recreate it. Thank you!
